# Cvičení

## Závod

Uvažuj časy závodníků za ročníky půlmaratonu 2019 a 2020, které jsou uloženy v souboru [half_marathon.csv](half_marathon.csv). V souboru je uloženo jméno závodníka, jeho čas a rok závodu, ke kterému se čas vztahuje. Tvým úkolem je spočítat, o kolik se změnil průměrný čas závodníků a zda se v průměru zlepšili či zhoršili (například protože kvůli lockdownům méně či více trénovali).



In [19]:
import pandas
df = pandas.read_csv("half_marathon.csv")
df.head()

Jmeno Cas zavodnika  Rok zavodu
0   Eisner Tomáš      01:21:25        2019
1  Hofhans Tomáš      01:27:43        2019
2        Krs Jan      01:28:23        2019
3      Kasal Jan      01:29:09        2019
4  Bryan Lauwers      01:30:18        2019

Můžeš využít následující postup:

- Seřaď si data (je třeba seřadit data dle jména závodníka i dle roku závodu).


In [20]:
df = df.sort_values(["Jmeno", "Rok zavodu"])
df.head()

Jmeno Cas zavodnika  Rok zavodu
35   Adamcová Helena      01:46:07        2019
76    Aster Vladimír      01:53:25        2019
374   Aster Vladimír      02:00:19        2020
231    Asterová Jana      02:16:59        2019
521    Asterová Jana      02:30:01        2020

- Převeď sloupec s časem závodníka na typ `datetime`. Můžeš využít stejný postup, jaký jsme si ukázali v lekci. Protože jde pouze o časový údaj, pandas k němu připojí dnešní datum, aby byly ve sloupci datum i čas. Toho si ale nevšímej, u obou sloupců je datum stejný, takže na porovnání údajů to nebude mít vliv.


In [22]:
df["Cas zavodnika"] = pandas.to_datetime(df["Cas zavodnika"])
df.head()

Jmeno       Cas zavodnika  Rok zavodu
35   Adamcová Helena 2023-04-20 01:46:07        2019
76    Aster Vladimír 2023-04-20 01:53:25        2019
374   Aster Vladimír 2023-04-20 02:00:19        2020
231    Asterová Jana 2023-04-20 02:16:59        2019
521    Asterová Jana 2023-04-20 02:30:01        2020

- Pomocí metody `shift()` si dej na jeden řádek výsledky obou závodů. Je nutné ji použít v kombinaci s metodou `groupby()`, jak je vidět níže. Je třeba nahradit `X` vhodně zvoleným číslem.

```py
df_joined["Cas zavodnika 2020"] = df_joined.groupby("Jmeno")["Cas zavodnika"].shift(X)
```

In [24]:
df["Cas zavodnika 2020"] = df.groupby("Jmeno")["Cas zavodnika"].shift(-1)
df.head()

Jmeno       Cas zavodnika  Rok zavodu  Cas zavodnika 2020
35   Adamcová Helena 2023-04-20 01:46:07        2019                 NaT
76    Aster Vladimír 2023-04-20 01:53:25        2019 2023-04-20 02:00:19
374   Aster Vladimír 2023-04-20 02:00:19        2020                 NaT
231    Asterová Jana 2023-04-20 02:16:59        2019 2023-04-20 02:30:01
521    Asterová Jana 2023-04-20 02:30:01        2020                 NaT


- Vyřaď z tabulky neúplné řádky pomocí metody `dropna()`.


In [26]:
df = df.dropna()
df.head()

Jmeno       Cas zavodnika  Rok zavodu  Cas zavodnika 2020  \
76       Aster Vladimír 2023-04-20 01:53:25        2019 2023-04-20 02:00:19   
231       Asterová Jana 2023-04-20 02:16:59        2019 2023-04-20 02:30:01   
101       Baborová Anna 2023-04-20 01:58:20        2019 2023-04-20 01:52:25   
132          Bambas Jan 2023-04-20 02:02:59        2019 2023-04-20 02:35:39   
324  Barochovská Andrea 2023-04-20 03:01:25        2019 2023-04-20 02:50:05   

                Zmena  
76    0 days 00:06:54  
231   0 days 00:13:02  
101 -1 days +23:54:05  
132   0 days 00:32:40  
324 -1 days +23:48:40

- Vypočítej rozdíl mezi časy závodníka a převeď ho na sekundy (postup jsme si ukazovali v lekci). Dále spočítej průměrnou změnu. Vyšlo i kladné nebo záporné číslo? A co to znamená.


In [29]:
df["Zmena"] = df["Cas zavodnika 2020"] - df["Cas zavodnika"]

df["Zmena v sekundach"] = df["Zmena"].dt.total_seconds()
df["Zmena v sekundach"].mean()

df.head()

583.7731958762887

### Výsledek

Mělo by ti vyjít, že průměrná změna je asi 584 sekund. To, jestli je číslo kladné nebo záporné záleží na výpočtu rozdílu. Při výpočtu `df["Cas zavodnika 2020"] - df["Cas zavodnika"]` by číslo mělo být kladné a znamená zhoršení (v roce 2020 závodníci běželi déle, čili se zhoršili). Při výpočtu `df["Cas zavodnika"] - df["Cas zavodnika 2020"]` pak vyjde rozdíl záporný (v roce 2019 běželi závodníci kratší čas, proto došlo ke zhoršení).



### Bonus

- Rozděl pomocí funkce `cut()` závodníky na ty, kteří se zlepšili a ty, kteří se zhoršili. Jako hranice můžeš využít "mínus nekonečno" (`float("-inf")`) a "plus nekonečno" (`float("inf")`). Dále pomocí agregace zjisti počty závodníků, kteří se zlepšili a kteří se naopak zhoršili

In [44]:
# Bonus
df["Zmena Text"] = pandas.cut(df["Zmena v sekundach"], bins=[float("-inf"), 0 , float("inf")], labels=["zlepsil se", "zhorsil se"])
df.head()

Jmeno       Cas zavodnika  Rok zavodu  Cas zavodnika 2020  \
76       Aster Vladimír 2023-04-20 01:53:25        2019 2023-04-20 02:00:19   
231       Asterová Jana 2023-04-20 02:16:59        2019 2023-04-20 02:30:01   
101       Baborová Anna 2023-04-20 01:58:20        2019 2023-04-20 01:52:25   
132          Bambas Jan 2023-04-20 02:02:59        2019 2023-04-20 02:35:39   
324  Barochovská Andrea 2023-04-20 03:01:25        2019 2023-04-20 02:50:05   

                Zmena  Zmena v sekundach  Zmena Text  
76    0 days 00:06:54              414.0  zhorsil se  
231   0 days 00:13:02              782.0  zhorsil se  
101 -1 days +23:54:05             -355.0  zlepsil se  
132   0 days 00:32:40             1960.0  zhorsil se  
324 -1 days +23:48:40             -680.0  zlepsil se

Dále pomocí agregace zjisti počty závodníků, kteří se zlepšili a kteří se naopak zhoršili.

In [46]:
df.groupby("Zmena Text").size()


Zmena Text
zlepsil se    11
zhorsil se    86
dtype: int64